cpp/test/distance/sparse_l2_nn.cu

```diff
@@ -0,0 +1,494 @@
+  // scan of the group lengths)
+  //
+  // - The first group always starts at index zero, so we do not store it.
+  //
+  // - The group_idxs[num_groups - 1] should always equal n.
+
+  if (blockIdx.y == 0 && threadIdx.y == 0) {
+    const int j_stride = blockDim.x * gridDim.x;
+    for (int j = blockIdx.x * blockDim.x + threadIdx.x; j < num_groups; j += j_stride) {
+      group_idxs[j] = (j + 1) * (n / num_groups);
+    }
+    group_idxs[num_groups - 1] = n;
+  }
+}
+
+template <typename DataT, bool Sqrt>
+class SparseL2NNTest : public ::testing::TestWithParam<Inputs<DataT>> {
+ public:
+  SparseL2NNTest()
+    : params(::testing::TestWithParam<Inputs<DataT>>::GetParam()),
+      stream(handle.get_stream()),
+      x(params.m * params.k, stream),
+      y(params.n * params.k, stream),
+      adj(params.m * params.num_groups, stream),
+      group_idxs(params.num_groups, stream),
+      xn(params.m, stream),
+      yn(params.n, stream),
+      min(params.m, stream),
+      min_ref(params.m, stream),
+      workspace(params.m * sizeof(int), stream)
+  {
+  }
+
+ protected:
+  void SetUp() override
+  {
+    raft::random::RngState r(params.seed);
+    int m          = params.m;
+    int n          = params.n;
+    int k          = params.k;
+    int num_groups = params.num_groups;
+    uniform(handle, r, x.data(), m * k, DataT(-1.0), DataT(1.0));
+    uniform(handle, r, y.data(), n * k, DataT(-1.0), DataT(1.0));
+
+    dim3 block(32, 32);
+    dim3 grid(10, 10);
+    init_adj<<<grid, block, 0, stream>>>(
+      m, n, num_groups, params.pattern, adj.data(), group_idxs.data());
+    RAFT_CUDA_TRY(cudaGetLastError());
+
+    generateGoldenResult();
+    raft::linalg::rowNorm(xn.data(), x.data(), k, m, raft::linalg::L2Norm, true, stream);
+    raft::linalg::rowNorm(yn.data(), y.data(), k, n, raft::linalg::L2Norm, true, stream);
+    RAFT_CUDA_TRY(cudaStreamSynchronize(stream));
+  }
+
+ protected:
+  Inputs<DataT> params;
+  rmm::device_uvector<DataT> x;
+  rmm::device_uvector<DataT> y;
+  rmm::device_uvector<bool> adj;
+  rmm::device_uvector<int> group_idxs;
+  rmm::device_uvector<DataT> xn;
+  rmm::device_uvector<DataT> yn;
+  rmm::device_uvector<cub::KeyValuePair<int, DataT>> min;
+  rmm::device_uvector<cub::KeyValuePair<int, DataT>> min_ref;
+  rmm::device_uvector<char> workspace;
+  raft::handle_t handle;
+  cudaStream_t stream;
+
+  virtual void generateGoldenResult()
+  {
+    int m          = params.m;
+    int n          = params.n;
+    int k          = params.k;
+    int num_groups = params.num_groups;
+
+    naive<DataT, Sqrt>(min_ref.data(),
+                       x.data(),
+                       y.data(),
+                       adj.data(),
+                       group_idxs.data(),
+                       m,
+                       n,
+                       k,
+                       num_groups,
+                       (int*)workspace.data(),
+                       stream);
+  }
+
+  void runTest(cub::KeyValuePair<int, DataT>* out)
```

Comment:
Can we use `raft::KeyValuePair` here instead? This is bound to cause an issue when, for example, someone adds an explicit template specialization which now forces symbols in the `libraft` binary to include `cub` types. We have encountered this issue before and it caused a great deal of symbols conflicts for our users).